Show entry background color as column

newhinton opened the pull request:
This PR adds a setting: 'Show Background Color of Entry' which disables the background color of a row if an entry has one. 
It also introduces a new Column: 'Entrycolor' which can show the disabled background color as a box, similar to the screenshot in #6553

Todo: 
~~I would like to change the column-name to an icon, a paintbrush or similar.~~ Done
~~I also would like to move the color-column infront of the entry-title~~ Done

Fixes #6553

## Type of change
- ✅ New feature (change that adds functionality)

droidmonkey commented:
Ehhhhhh I'm really not a fan of the setting or the additional column. Who is that setting targeted to? People who actually use the background feature wouldn't use it, they explicitly set their backgrounds. The better fix for this is to properly display the foreground text color depending on the current background color.

newhinton commented:
I generally think that the "whole-row" coloring is quite obstrusive. A more toned-down version would be great, and some applications do it by coloring the icons. Providing a column that displays the color in a small box would be a good way to do it, while not interfering with theming and the contrasts. 

I agree, maybe it would be better to make the background conditional, and remove the checkbox. If the color-row is shown, the background is not applied, but if it is hidden, it is colored. I have to find out how to do it though.

newhinton commented:
I have now made the setting implicit. If the user shows the color-column, the row is not colored anymore. If it is hidden, the row is colored as usual.
An additional bonus: The color-column is now sortable.

newhinton commented:
The old list:

![grafik](https://user-images.githubusercontent.com/25279821/127904104-a2f980a7-aecd-4273-9546-ee8caeea8e1c.png)

The new list, with enabled color-column:
![grafik](https://user-images.githubusercontent.com/25279821/127904339-b26a38d7-c751-4a74-916c-05b4d51baaf4.png)


I think the latter one is a whole lot more readable, and more consistent with the overall look of the application.

I guess it is not possible to preserve a custom order when the enumeration is changed?

droidmonkey commented:
That's a great compromise. How are you sorting the colors? Could do it by the total integer value of the color.

newhinton commented:
I think the color is sorted by its value alphabetical, it is basically an unintended consequence. Therefore while it is beeing sorted, its not really how humans would sort it visually

Edit:
I have added an icon for the row, which i find more fitting. I took it from the material design repo, and i hope i added all the required notes to it

newhinton commented:
I am currently trying to fix the original issue aswell, but i have trouble determining wether a color is a light one or not, is there a util for this?


Edit:

I have added code that tries to calculate luminosity based on this post:
https://stackoverflow.com/a/596243
The threshhold may need to be adjusted

droidmonkey commented:
This function should give you a "brightness" score:https://doc.qt.io/qt-5/qcolor.html#lightness

droidmonkey commented on src/gui/entry/EntryModel.cpp:
```diff
@@ -20,7 +20,9 @@
 #include <QFont>
 #include <QMimeData>
 #include <QPalette>
+#include <zconf.h>
```
What is this for?

newhinton replied:
Good Question. Must have slipped in from me experimenting, i removed it

droidmonkey commented on src/gui/entry/EntryModel.h:
```diff
@@ -34,20 +34,21 @@ class EntryModel : public QAbstractTableModel
-        Totp = 12,
-        Size = 13,
-        PasswordStrength = 14
+        Color = 1,
```
This should be at the bottom not the top

newhinton replied:
So it should be ID 15 instead of 1? 

I assumed the default order was defined here, and i put it where i thought best of, logically.

droidmonkey commented:
You need to run `make format`

phoerious commented:
I don't like this extra column with its big title. I'd rather put a small block at the start of the line.

newhinton commented:
@phoerious i already exchanged the title with an icon, so that the width matches the box

phoerious commented:
I would still prefer just a small colour swatch at the beginning of the line without a separate header. The horizontal space is precious.

newhinton commented:
what do you mean with 'colour swatch'?

this implementation allows the user to switch between a colored box and the full row beeing colored. If a box would take to much space, you can choose to not enable it. (and therefore keep it like it is currently)

Also, the box should only take as much width as the box takes, the header should be smaller

I hope i could clarify how it is supposed to work, if you still have concerns maybe we can solve them

droidmonkey commented:
I'm ok with the current implementation

michaelk83 commented:
> i already exchanged the title with an icon, so that the width matches the box

> I hope i could clarify how it is supposed to work

@newhinton if you can update the screenshot to show the icon in the column title, it would be clearer for everyone.

phoerious commented:
Something like this:

![image](https://user-images.githubusercontent.com/911270/135730853-6f8c9f81-2551-4b2b-bf27-871a75120fdd.png)

And I don't like the variant where the entire row is coloured at all. It's ugly, it doesn't work well with any theme, and it doesn't even meet minimal contrast requirements.

newhinton commented:
![grafik](https://user-images.githubusercontent.com/25279821/135748767-99d57b65-9c72-483e-9695-7b709a2a8d22.png)

This is how it currently looks when the Backgroundcolor as box is selected in the columnselection

phoerious commented:
It's too wide and colour may be a useful filter criterion, but not a very useful sorting criterion.

newhinton commented:
Any news on this?

The last comment was that its too wide, but that is just how qt handles textless columns. Attachment and TOTP behaves the same way.

droidmonkey commented:
I don't really understand the purpose of choosing colors for fg/bg... it may make sense for KeePass2 but not for us. No other ports care about this feature. I'm frankly in favor of dropping support entirely in 2.8.0 now that we have robust tags support.

newhinton commented:
> I'm frankly in favor of dropping support entirely in 2.8.0 now that we have robust tags support.

That is totally also a solution. Dropping those colors is also an option, it just is that the current way is bad and needs some change. Either remove it altogether or make changes should fix it.

Has there been surveys of the usage of those colors?

droidmonkey commented:
@phoerious what is your position, I mean it is user choice to begin with. You can't search based on color, you can't sort based on color, all it does is presentation.

newhinton commented:
> you can't sort based on color,

That is not _entirely_ true. You actually can sort by color, and it does order them. The ordering is just not what one expects, but it does group colors if you sort by that columb.


If it is a useful sorting, that's entirely a different thing, and probably highly subjective.

newhinton commented:
@droidmonkey Any news on this? 

I totally can update this PR, but for that, i'd like to have an idea if this is going to be accepted at some point. I am also open to any changes or suggestions.

droidmonkey commented:
I cannot spend time on this one and I think if we can't get this to be like the image that @phoerious posted above then just cancel it altogether. With the introduction of tags and saved searches, I don't see foreground/background as being a feature I even want to support at all.

newhinton commented:
Okay, so if i can make it look like his screenshot, this will be seriously considered for merging?

If so, i will try to make it that way. I cannot promise that it will look *exactly* the same, but i will try to get as close as possible. I think removing the reorder-button and the icon from the column-header should get me very close.

newhinton commented:
![image](https://github.com/keepassxreboot/keepassxc/assets/25279821/f2090513-589f-4f96-940d-387a7f7c8199)


This is the smallest i can make it.*

The first box is the icon field, but with a background color set.
The second and third, are unicode-chars, that are colored as text.

The first box is the least "complicated", it does not require workarounds, but it is rather big.
The other ones are smaller, and more customizable (for us, the nicest looking one should be used, any unicode-character that supports text-styling should be working)
The drawback: as you can see in the screenshot, the third row is selected, and it looses its styling. This might be fixable, but it might require more code than it would be worth it.

I also will remove all the code for the old styling. That means the background color will not be rendered anymore, unless the user enables the specific column for it. 


*It is currently bigger than it needs to be because i enabled all three boxes. However, it will not be smaller than the paperclip-column. It seems it is impossible to make it smaller., but i will have a look again tomorrow.

![image](https://github.com/keepassxreboot/keepassxc/assets/25279821/291102a3-4cd2-4ee2-9557-71709dd7cb1b)
The fourth option: limit rendering of background color to a single empty cell. That one requires no workaround, but i would prefer the first box if i cant make it smaller.

Let me know what you think. I really do think that we need to so something, because colors from other keepass-managers simply dont work at all with the current implementation.

newhinton commented:
![image](https://github.com/keepassxreboot/keepassxc/assets/25279821/750c9a05-79dd-4a1e-8951-314f2294e66e)

Okay, got it. This removes the minimum restriction for the color-column.

droidmonkey commented:
That last screenshot is very close and promising.

newhinton commented:
> That last screenshot is very close and promising.

What changes do you want?
That one uses the color as icon, any styling is done by qt itself. It could be made smaller, but i'd advise against that because it makes rearranging very hard. It still is "just" a column.  (the same goes for fixing it in place)

droidmonkey commented:
It could be confusing compared to the password strength column, which uses the same square. However, I guess it doesn't matter since the end user is responsible for enabling and using this feature. Perhaps we could use a skinny rectangle instead of a square or a circle? Just dunno how that would look. 

To be honest I don't really like the password strength square either but that's not up for change in this pr.

newhinton commented:
![image](https://github.com/keepassxreboot/keepassxc/assets/25279821/1033126b-03a3-43b1-8cd9-f507eaf1b0fa)

![image](https://github.com/keepassxreboot/keepassxc/assets/25279821/8d7dd596-cec0-421e-8406-3974151d1dee)

![image](https://github.com/keepassxreboot/keepassxc/assets/25279821/3b021045-4a1b-4ae3-a778-60f86962f09a)


Which one do you like?

droidmonkey commented:
Oh, the middle one!

newhinton commented:
Done! Do you want me to take a look at the password-strength indicator aswell? (In a different PR)

droidmonkey commented:
Can probably just leave that thing alone, but if you want to mock out a new look then I won't stop you 😉

newhinton commented:
Now that i figured out how to make it tiny, it was rather easy:

![image](https://github.com/keepassxreboot/keepassxc/assets/25279821/7fc2af23-c91a-47e2-a11c-acc529704b2f)

![image](https://github.com/keepassxreboot/keepassxc/assets/25279821/efb85199-20af-4c29-b9f2-6170f038bdf1)

Though the same applies here: if the row is selected, the color of the text will change back to the default.

Edit:

uhhhh
![image](https://github.com/keepassxreboot/keepassxc/assets/25279821/a703fa21-f519-4109-91ef-2832de5af628)

Edit 2:
![image](https://github.com/keepassxreboot/keepassxc/assets/25279821/27ba2952-1502-4a76-8718-e4d30199db1f)

droidmonkey commented on src/gui/entry/EntryModel.cpp:
```diff
@@ -414,6 +433,8 @@ QVariant EntryModel::headerData(int section, Qt::Orientation orientation, int ro
         case Totp:
             return tr("Has TOTP");
+        case Color:
+            return tr("Show Backgroundcolor as Box");
```
Recommend just labeling this "Background Color"

droidmonkey commented on src/gui/entry/EntryModel.h:
```diff
@@ -85,6 +87,7 @@ private slots:
     void severConnections();
     void makeConnections(const Group* group);
 
+    bool m_colorVisible = true;
```
This variable (and corresponding function) is ambiguous. Should add the word "Background" in both.

droidmonkey commented on src/gui/entry/EntryView.cpp:
```diff
@@ -474,6 +480,8 @@ void EntryView::resetViewToDefaults()
     header()->hideSection(EntryModel::Attachments);
     header()->hideSection(EntryModel::Size);
     header()->hideSection(EntryModel::PasswordStrength);
+    header()->hideSection(EntryModel::Color);
+    m_model->setColorVisible(true);
```
Instead of hard coding the state change in various spots in the code, recommend adding a private function such as "onHeaderChanged" and then do the generic check/update in that function. Call that function from all the places you are setting the color visible. `m_model->setColorVisible(isColumnHidden(EntryModel::Color));`

newhinton replied:
Did you mean it like i applied it? It was just called in two places, but its more reusable now.

droidmonkey replied:
Yah we might expand on this function in the future. I noticed there is no signal from QHeaderView for when a column is hidden/visible changed.

newhinton committed:
Show Row-Backgroundcolor in a column

Fixes #6553

Allow users to choose to show the entry background color as a column instead of changing the background of the entire row.
